freud/diffraction.pyx

```diff
@@ -0,0 +1,313 @@
+             [ss[1, 1], ss[0, 1], roll_shift],
+             [0, 0, 1]])
+
+        zoom_matrix = np.diag((self.zoom, self.zoom, 1))
+
+        # This matrix uses homogeneous coordinates. It is a 3x3 matrix that
+        # transforms 2D points and adds an offset.
+        inverse_transform = np.linalg.inv(
+            zoom_matrix @ shear_matrix @ shift_matrix)
+
+        img = scipy.ndimage.affine_transform(
+            input=img,
+            matrix=inverse_transform,
+            output_shape=(self.output_size, self.output_size),
+            order=1,
+            mode="constant")
+        return img
+
+    def compute(self, system, view_orientation=None):
+        R"""Computes diffraction pattern.
+
+        Args:
+            system:
+                Any object that is a valid argument to
+                :class:`freud.locality.NeighborQuery.from_system`.
+            view_orientation ((:math:`4`) :class:`numpy.ndarray`, optional):
+                View orientation. Uses :math:`(1, 0, 0, 0)` if not provided
+                or :code:`None` (Default value = :code:`None`).
+        """
+        system = freud.locality.NeighborQuery.from_system(system)
+
+        if view_orientation is None:
+            view_orientation = np.array([1., 0., 0., 0.])
+
+        grid_size = int(self.grid_size / self.zoom)
+
+        # Compute the box projection matrix
+        inv_shear = self._calc_proj(view_orientation, system.box)
+
+        # Rotate points by the view quaternion and shear by the box projection
+        xy = rowan.rotate(view_orientation, system.points)[:, 0:2]
+        xy = xy @ inv_shear.T
+
+        # Map positions to [0, 1] and compute a histogram "image"
+        xy += 0.5
+        xy %= 1
+        im, _, _ = np.histogram2d(
+            xy[:, 0], xy[:, 1], bins=np.linspace(0, 1, grid_size))
+
+        # Compute FFT and convolve with Gaussian
+        cdef double complex[:, :] diffraction_fft
+        diffraction_fft = np.fft.fft2(im)
+        diffraction_fft = scipy.ndimage.fourier.fourier_gaussian(
+            diffraction_fft, self.peak_width / self.zoom)
+        diffraction_fft = np.fft.fftshift(diffraction_fft)
+
+        # Compute the squared modulus of the FFT, which is S(q)
+        self._diffraction = np.real(
+            diffraction_fft * np.conjugate(diffraction_fft))
+
+        # Transform the image (scale, shear, zoom) and normalize S(q) by N^2
+        N = len(system.points)
+        self._diffraction = self._transform(
+            self._diffraction, system.box, inv_shear) / (N*N)
+
+        # Compute a cached array of k-vectors that can be rotated and scaled
+        if not self._called_compute:
+            # Create a 1D axis of k-vector magnitudes
+            self._k_values_orig = np.fft.fftshift(np.fft.fftfreq(
+                n=self.output_size,
+                d=1/self.output_size))
+
+            # Create a 3D meshgrid of k-vectors with shape
+            # (output_size, output_size, 3)
```

Comment:
```suggestion
            # (grid_size, grid_size, 3)
```

Reply:
I think this suggestion is incorrect. Please verify.

Reply:
Yes, if I had known these arrays were initialized incorrectly, I would not have made this suggestion.